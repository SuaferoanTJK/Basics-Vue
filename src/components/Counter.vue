<script>
export default {
  props: {
    title: String,
    start: {
      type: Number,
      default: 100,
      validator(value) {
        return value > 0 ? true : false;
      },
    },
  },
  name: "Counter",
  data() {
    return {
      counter: this.start,
      exp: 3,
    };
  },
  methods: {
    increase() {
      this.counter++;
    },
    decrease() {
      this.counter--;
    },
  },
  computed: {
    squareCounted() {
      return Math.pow(this.counter, this.exp);
    },
    customTitle() {
      return this.title ? this.title : "Counter";
    },
  },
};
</script>

<template>
  <img class="w-80" src="../assets/logo.png" alt="Vue logo" />
  <h2 data-test="title" class="text-3xl font-bold text-white">
    {{ customTitle }}
  </h2>
  <p data-test="counter" class="text-xl text-white">
    {{ counter }} <sup>{{ exp }}</sup> = {{ squareCounted }}
  </p>
  <div class="flex gap-2 mt-2">
    <button
      @click="decrease"
      class="rounded bg-slate-300 hover:bg-slate-400 p-2"
    >
      Decrease
    </button>
    <button
      @click="increase"
      class="rounded bg-slate-300 hover:bg-slate-400 p-2"
    >
      Increase
    </button>
  </div>
</template>
