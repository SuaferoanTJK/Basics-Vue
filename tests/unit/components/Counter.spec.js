import { shallowMount } from "@vue/test-utils";
import Counter from "@/components/Counter";
import { findByTestAttr } from "../utils";

describe("Counter Unit Testing", () => {
  let component;
  beforeAll(() => {
    component = shallowMount(Counter);
  });

  test("snapshot should matched", () => {
    expect(component.html()).toMatchSnapshot();
  });
  test("customTitle should have the default value of 'Counter'", () => {
    const customTitle = findByTestAttr(component, "title").text();
    expect(customTitle).toBe("Counter");
  });
  test("exp should have the default value of '3'", () => {
    const exp = findByTestAttr(component, "counter").text().split(" ")[1];
    expect(exp).toBe("3");
  });
  test("counter should have the default value of '100'", () => {
    const counter = findByTestAttr(component, "counter").text().split(" ")[0];
    expect(counter).toBe("100");
  });
  test("counter must increase or decrease by 1", async () => {
    const wrapper = shallowMount(Counter);
    const [decreaseBtn, increaseBtn] = wrapper.findAll("button");
    await increaseBtn.trigger("click");
    await decreaseBtn.trigger("click");
    await decreaseBtn.trigger("click");
    await decreaseBtn.trigger("click");
    const counter = findByTestAttr(wrapper, "counter").text().split(" ")[0];
    expect(counter).toBe("98");
  });
  test("must establish the default value for prop counter", () => {
    const { start } = component.props();
    const counter = findByTestAttr(component, "counter").text().split(" ")[0];
    expect(Number(counter)).toBe(start);
  });
  test("should display the prop title", () => {
    const title = "Hi World";
    const component = shallowMount(Counter, { props: { title } });
    const customTitle = findByTestAttr(component, "title").text();
    expect(customTitle).toBe(title);
  });
});
