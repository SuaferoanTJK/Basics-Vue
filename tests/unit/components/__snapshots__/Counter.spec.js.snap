// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Unit Testing snapshot should matched 1`] = `
<img class="w-80" src="" alt="Vue logo">
<h2 data-test="title" class="text-3xl font-bold text-white">Counter</h2>
<p data-test="counter" class="text-xl text-white">100 <sup>3</sup> = 1000000</p>
<div class="flex gap-2 mt-2"><button class="rounded bg-slate-300 hover:bg-slate-400 p-2"> Decrease </button><button class="rounded bg-slate-300 hover:bg-slate-400 p-2"> Increase </button></div>
`;
